File: shortest path in DAG.py
class Solution:
    def topoSort(self, currentNode, visited, adjList, stack):
        visited[currentNode] = 1
        for adjacentNode in adjList[currentNode]:
            v, _ = adjacentNode
            if not visited[v]:
                self.topoSort(v, visited, adjList, stack)
        stack.append(currentNode)
        
    def shortestPath(self, n : int, m : int, edges : List[List[int]]) -> List[int]:
        adjList = [[] for _ in range(n)]
        for current in edges:
            u, v, wt = current
            adjList[u].append((v, wt))
        stack = []
        visited = [0 for _ in range(n)]
        for ind in range(n):
            if not visited[ind]:
                self.topoSort(ind, visited, adjList, stack)
            
        distance = [float('inf') for _ in range(n)]
        distance[0] = 0
        while stack:
            currentNode = stack.pop()
            for adjacentNode in adjList[currentNode]:
                v, wt = adjacentNode
                if distance[currentNode] + wt < distance[v]:
                    distance[v] = distance[currentNode] + wt
        for index, value in enumerate(distance):
            if value == float('inf'):
                distance[index] = -1
        return distance
            

File: shortest path in DAG.cpp
class Solution {
public:
    void dfs(int current, vector<vector<pair<int, int>>>& adjList, vector<int>& visited, vector<int>& nodes) {
        visited[current] = 1;
        for (auto& it : adjList[current]) {
            int adjacentNode = it.first;
            if (!visited[adjacentNode]) {
                dfs(adjacentNode, adjList, visited, nodes);
            }
        }
        nodes.push_back(current);
    }
    vector<int> shortestPath(int N, int M, vector<vector<int>>& edges) {
        vector<vector<pair<int, int>>> adjList(N);
        for (auto& it : edges) {
            int u = it[0];
            int v = it[1];
            int wt = it[2];
            adjList[u].push_back({v, wt});
        }
        vector<int> topoSort;
        vector<int> visited(N, 0);
        
        for (int i = 0; i < N; i++) {
            if (!visited[i]) {
                dfs(i, adjList, visited, topoSort);
            }
        }
        vector<int> distance(N, 1e9);
        distance[topSort.back() - 1] = 0;
        while (topoSort.size() > 0) {
            int current = topoSort.back();
            topoSort.pop_back();
            // explore all neighbour nodes
            for (auto& adjacentNode : adjList[current]) {
                
                int v = adjacentNode.first;
                int wt = adjacentNode.second;
                if (distance[current] + wt < distance[v]){
                    
                    distance[v] = distance[current] + wt;
                }
            }
        }
        for (int i =0; i < N; i++){
            if (distance[i] == 1e9){
                distance[i] = -1;
            }
        }
        return distance;
            
    }
};
// time: O( V + E)
